dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -67,6 +83,52 @@ class TypedDataset[T](
     ): GroupedByManyOps[T, K, Out0] = new GroupedByManyOps[T, K, Out0](self, groupedBy)
   }
 
+  def join[A: TypedEncoder, B](
+    right: TypedDataset[A],
+    leftCol: TypedColumn[T, B],
```

Comment:
I was thinking about something like this, but then I've realized that 99% of joins use equality, and it's better to start with the simple solution.